Makefile

```diff
@@ -86,80 +168,99 @@ run: generate fmt vet manifests export_env_vars
 	envsubst < makefileutils.yaml | kubectl apply -f -
 	go run ./main.go
 
-# Install CRDs into a cluster
-install: manifests kustomize
+.PHONY: docker-build
+docker-build: test ## Build docker image with the manager.
+	docker build . -t ${IMG}
+
+.PHONY: docker-push
+docker-push: ## Push docker image with the manager.
+	docker push ${IMG}
+
+##@ Deployment
+
+.PHONY: install
+install: manifests kustomize ## Install CRDs into the K8s cluster specified in ~/.kube/config.
 	$(KUSTOMIZE) build config/crd | kubectl apply -f -
 	./shim/shim.sh install
 
-# Uninstall CRDs from a cluster
-uninstall: manifests kustomize
-	$(KUSTOMIZE) build config/crd | kubectl delete -f -
+.PHONY: uninstall
+uninstall: manifests kustomize ## Uninstall CRDs from the K8s cluster specified in ~/.kube/config. Call with ignore-not-found=true to ignore resource not found errors during deletion.
+	$(KUSTOMIZE) build config/crd | kubectl delete --ignore-not-found=$(ignore-not-found) -f -
 	./shim/shim.sh uninstall
 
-# Deploy controller in the configured Kubernetes cluster in ~/.kube/config
-deploy: manifests kustomize
+.PHONY: deploy
+deploy: manifests kustomize ## Deploy controller to the K8s cluster specified in ~/.kube/config.
 	cd config/manager && $(KUSTOMIZE) edit set image controller=${IMG}
 	$(KUSTOMIZE) build config/default | kubectl apply -f -
 
-# Generate manifests e.g. CRD, RBAC etc.
-manifests: controller-gen
-	$(CONTROLLER_GEN) $(CRD_OPTIONS) rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
+.PHONY: undeploy
+undeploy: ## Undeploy controller from the K8s cluster specified in ~/.kube/config. Call with ignore-not-found=true to ignore resource not found errors during deletion.
+	$(KUSTOMIZE) build config/default | kubectl delete --ignore-not-found=$(ignore-not-found) -f -
 
-# Run go fmt against code
-fmt:
-	go fmt ./...
+.PHONY: bundle
+bundle: manifests kustomize operator-sdk ## Generate bundle manifests and metadata, then validate generated files.
+	$(OPERATOR_SDK) generate kustomize manifests --interactive=false -q
+	cd config/manager && $(KUSTOMIZE) edit set image controller=$(IMG)
+	cd config/manifests/bases && \
+		rm -rf kustomization.yaml && \
+		$(KUSTOMIZE) create --resources ocs-osd-deployer.clusterserviceversion.yaml && \
+		$(KUSTOMIZE) edit add annotation --force 'olm.skipRange':">=0.0.1 <$(VERSION)" && \
+		$(KUSTOMIZE) edit add patch --name ocs-osd-deployer.v0.0.0 --kind ClusterServiceVersion \
+		--patch '[{"op": "replace", "path": "/spec/replaces", "value": "ocs-osd-deployer.v$(REPLACES)"}]'
+	$(KUSTOMIZE) build config/manifests | $(OPERATOR_SDK) generate bundle \
+		-q \
+		--extra-service-accounts prometheus-k8s \
+		--overwrite \
+		--version $(VERSION) \
+		$(BUNDLE_METADATA_OPTS) \
+		--output-dir=$(OUTPUT_DIR) \
+		$(BUNDLE_EXTRA_FLAGS)
```

Comment:
Why do we need both `BUNDLE_METADATA_OPTS` and the new `BUNDLE_EXTRA_FLAGS? can't we just use the original BUNDLE_METADATA_OPTS for all the customer-provided options?

Reply:
- This is due to below
``` makefile
# USE_IMAGE_DIGESTS defines if images are resolved via tags or digests
USE_IMAGE_DIGESTS ?= false
ifeq ($(USE_IMAGE_DIGESTS), true)
    BUNDLE_EXTRA_FLAGS += --use-image-digests
endif
```

Reply:
ooh! can be done. will update.

Reply:
done.